src/pointer.rs

```diff
@@ -315,30 +330,26 @@ impl Pointer {
     }
 
     /// Finds the commonality between this and another `Pointer`.
-    pub fn intersection<'a>(&'a self, other: &Self) -> &'a Self {
-        if &self.0 == "/" || &other.0 == "/" {
-            if self.0 != other.0 {
-                return Self::root();
-            } else {
-                return self;
-            }
+    pub fn intersection<'a>(&'a self, mut other: &Self) -> &'a Self {
+        if self.is_root() || other.is_root() {
+            return Self::root();
         }
-        let mut last_slash = 0;
         let mut idx = 0;
-        for (a, b) in self.0.bytes().zip(other.0.bytes()) {
-            if a != b {
-                return Self::new(&self.0[..last_slash]);
-            }
-            if a == b'/' {
-                last_slash = idx;
+        let mut this = self;
+        while let Some((tok, remaining)) = this.split_front() {
+            if let Some((other_tok, other_remaining)) = other.split_front() {
+                if tok != other_tok {
+                    break;
+                }
+                this = remaining;
+                other = other_remaining;
+                idx += tok.encoded().chars().count() + 1;
+            } else {
+                break;
             }
-            idx += 1;
-        }
-        if idx == self.0.len() || self.0.as_bytes()[idx] == b'/' {
-            Self::new(&self.0[..idx])
-        } else {
-            Self::new(&self.0[..last_slash])
         }
```

Comment:
I could be wrong, but wouldn't it be simpler to do something like `self.tokens().zip(other.tokens())`?

Also we should definitely add `Token::len`.

Reply:
Yea, it probably would be easier to use `tokens`.

I keep using `Chars::count` because in most cases, the tokens are looped over using chars... but then I screwed up and used byte index in `split_at` which may be why I'm seeing failing tests that aren't immediately obvious

```rust
    pub fn split_at(&self, idx: usize) -> Option<(&Self, &Self)> {
        if self.0.get(idx..idx + 1) != Some("/") {
            return None;
        }
        let (head, tail) = self.0.split_at(idx);
        Some((Self::new(head), Self::new(tail)))
    }
```

Reply:
No, I may be screwing up by using `Chars::count` altogether.

Reply:
I'm not sure what the right answer is - to continue using `Chars` or switch to `bytes` everywhere.

Reply:
That was it. Switching it to `bytes` flipped it green.

I'll add `len` to `Pointer` and `Token`.

Reply:
Actually.. adding `len` to `Token` is going to be confusing, right?

Reply:
I think it's fine, it's like how `String::len` returns the number of bytes, not chars, even though not every byte is displayed. We should document it clearly, but the semantics is consistent with `std`.